tabbycat/venues/allocator.py

```diff
@@ -168,7 +196,139 @@ def allocate_unconstrained_venues(self, debates):
         random.shuffle(debates)
         return {debate: venue for debate, venue in zip(debates, self._preferred_venues)}
 
-    def save_venues(self, debate_venues):
-        for debate, venue in debate_venues.items():
-            debate.venue = venue
-        Debate.objects.bulk_update(debate_venues.keys(), ['venue'])
+
+class BaseHungarianVenueAllocator(BaseVenueAllocator):
+    """Base class for venue allocations using the Hungarian algorithm.
+
+    Most of the costs would remain the same between allocators. Works as an
+    allocator on its own (no cost for history)"""
+
+    description = _("Hungarian")
+
+    def __init__(self, round, debates=None):
+        super().__init__(round, debates)
+
+        t = self.round.tournament
+        self.history_cost = t.pref('venue_history_cost')
+        self.constraint_cost = t.pref('venue_constraint_cost')
+        self.bad_venue_cost = t.pref('venue_score_cost')
+
+        self.munkres = Munkres()
+
+    def get_debate_queryset(self):
+        """For history constraints.
+
+        Get previous debates (with venues & categories) of participants in addition to current ones"""
+
+        return self.round.debate_set.all().select_related('venue').prefetch_related(
+            Prefetch('debateadjudicator_set', queryset=DebateAdjudicator.objects.select_related('adjudicator', 'adjudicator__institution')),
+            Prefetch('debateteam_set', queryset=DebateTeam.objects.select_related('team', 'team__institution')),
+        )
+
+    def allocate(self):
+        self.prev_venues = self.get_previous_venue_categories()
+        self.constraints = self.collect_dict_constraints()
+
+        self.max_venue_score = self.venues[0].priority
+        self.max_constraint_priority = list(self.constraints.values())[0][0].priority
```

Comment:
This line crashes with an `IndexError` if `self.constraints` is empty.

(This is the only "change requested".)